File: BookLibrary/Server/LibraryService.svc.fs
namespace BookLibrary.Server

open System
open System.Collections.Generic
open System.ServiceModel

type BookStorage() =
    let mutable _books : list<Book> = [
        new Book(
            Id = 1,
            Name = "The C Programming Language",
            Author = "Kernighan, Ritchie",
            PublishYear = 1975,
            Type = BookType.Other);
        new Book(
            Id = 2,
            Name = "Homotopy Type Theory: Univalent Foundations of Mathematics",
            Author = "The Univalent Foundations Program",
            PublishYear = 2013,
            Type = BookType.Scientific);
        new Book(
            Id = 3,
            Name = "Certified Programming with Dependent Types",
            Author = "Adam Chlipala",
            PublishYear = 2013,
            Type = BookType.Scientific);
        ]

    member storage.Books
        with get() = _books
        and set(value) = _books <- value

[<ServiceBehavior(InstanceContextMode = InstanceContextMode.PerSession)>]
type LibraryService() =
    static let _bookStorage : BookStorage = new BookStorage()
    let mutable _currentUser : Option<UserInfo> = Option.None
    let mutable _takenBooks : list<int> = list.Empty
    let mutable _returnedBooks : list<int> = list.Empty

    interface ILibraryService with
        member x.EnterLibrary id name =
            _currentUser <- Option.Some(new UserInfo(Id = id, Name = name))

        member x.AddBook book =
            _bookStorage.Books <- (book::_bookStorage.Books)

        member x.GetBook bookId =
            _bookStorage.Books
                |> List.tryFind (fun book -> book.Id = bookId )
                |> function None -> null | Some book -> book

        member x.GetBooks authorName =
            _bookStorage.Books
                |> List.filter (fun book -> book.Author = authorName)
                |> Seq.cast

        member x.TakeBook bookId =
            _bookStorage.Books 
                |> List.tryFind (fun book -> book.Id = bookId)
                |> function
                    | None -> failwith "Book was not found"
                    | Some book ->
                        match book.Taken with
                        | true -> failwith "Book is already taken"
                        | _ ->
                            match _takenBooks.Length with
                            | 5 -> failwith "You are too greedy asshole"
                            | _ -> _takenBooks <- (bookId :: _takenBooks)

        member x.ReturnBook bookId =
            _bookStorage.Books
                |> List.tryFind (fun book -> book.Id = bookId)
                |> function
                    | None -> failwith "Book was not found"
                    | Some book ->
                        match book.Taken, book.TakerInfo with
                        | true, _currentUser -> _returnedBooks <- (bookId :: _returnedBooks)
                        | _ -> failwith "Book was not taken by you"

        member x.SaveChanges () =
            match _currentUser with
                | Option.None -> failwith "You didn't authenticated"
                | Some userInfo ->
                    List.iter (fun bookId  ->
                        _bookStorage.Books
                            |> List.tryFind (fun b -> b.Id = bookId)
                            |> function
                                | None -> failwith "Book was not found"
                                | Some book -> book.Taken <- true
                                               book.TakerInfo <- Some userInfo) _takenBooks
                    List.iter (fun bookId  ->
                        _bookStorage.Books
                            |> List.tryFind (fun b -> b.Id = bookId)
                            |> function
                                | None -> failwith "Book was not found"
                                | Some book -> book.Taken <- false
                                               book.TakerInfo <- None) _returnedBooks

        member x.Leave () = ()

File: BookLibrary/Server/LibraryService.svc.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.ServiceModel;

namespace Server
{
    public class BookStorage
    {
        public List<Book> BookList = new List<Book>
        {
            new Book
            {
                Id = 1,
                Name = "The C Programming Language",
                Author = "Kernighan, Ritchie",
                PublishYear = 1975,
                Type = BookType.Other
            },
            new Book
            {
                Id = 2,
                Name = "Homotopy Type Theory: Univalent Foundations of Mathematics",
                Author = "The Univalent Foundations Program",
                PublishYear = 2013,
                Type = BookType.Scientific
            },
            new Book
            {
                Id = 3,
                Name = "Certified Programming with Dependent Types",
                Author = "Adam Chlipala",
                PublishYear = 2013,
                Type = BookType.Scientific
            }
        };
    }

    [ServiceBehavior(InstanceContextMode = InstanceContextMode.PerSession)]
    public class LibraryService : ILibraryService
    {
        private const int MaxBooksInOneHands = 5;

        private static readonly BookStorage Storage = new BookStorage();
        private UserInfo _currentUser;
        private readonly List<int> _returnedBooks = new List<int>();
        private readonly List<int> _takenBooks = new List<int>();
        private ISaveChangesOperationCallback Callback => OperationContext.Current.GetCallbackChannel<ISaveChangesOperationCallback>();

        public void AddBook(Book book)
        {
            Storage.BookList.Add(book);
        }

        public void EnterLibrary(int userId, string userName)
        {
            _currentUser = new UserInfo
            {
                Id = userId,
                Name = userName
            };
        }

        public Book GetBook(int bookId)
        {
            var book = Storage.BookList.FirstOrDefault(x => x.Id == bookId);
            if (book == null)
            {
                throw new FaultException("Книга не найдена");
            }
            return book;
        }

        public IEnumerable<Book> GetBooks(string authorName)
        {
            return Storage.BookList.Where(x => x.Author == authorName);
        }

        public void Leave() { }

        public void ReturnBook(int bookId)
        {
            var book = Storage.BookList.FirstOrDefault(x => x.Id == bookId);
            if (book == null)
            {
                throw new FaultException("Книга не найдена");
            }
            if (!book.Taken)
            {
                throw new FaultException("Книга не была ранее взята из библиотеки");
            }
            if (book.TakerInfo.Id != _currentUser.Id)
            {
                throw new FaultException("Не ты брал эту книгу из библиотеки");
            }
            _returnedBooks.Add(bookId);
        }

        public void SaveChanges()
        {
            foreach (var book in Storage.BookList.Where(x => _returnedBooks.Contains(x.Id)))
            {
                book.Taken = false;
                book.TakerInfo = null;
            }
            foreach (var book in Storage.BookList.Where(x => _takenBooks.Contains(x.Id)))
            {
                book.TakenDateTime = DateTime.Now;
                book.Taken = true;
                book.TakerInfo = _currentUser;
            }
            //if (Storage.BookList.Any(x => x.Taken && x.TakerInfo.Id == _currentUser.Id
            //    && (DateTime.Now - x.TakenDateTime) > TimeSpan.FromDays(30)))
            {
                Callback.OnSaveChanges();
            }
        }

        public void TakeBook(int bookId)
        {
            var book = Storage.BookList.FirstOrDefault(x => x.Id == bookId);
            if (book == null)
            {
                throw new FaultException("Книга не найдена");
            }
            if (book.Taken)
            {
                throw new FaultException("Эту книгу уже кто-то взял");
            }
            if (Storage.BookList.Count(x => x.Taken && x.TakerInfo.Id == _currentUser.Id) >= MaxBooksInOneHands)
            {
                throw new FaultException("У тебя и так дофига книг");
            }
            _takenBooks.Add(bookId);
        }
    }
}
